File: backend/internal/mapgen/classify.go
package mapgen

// Technique is the extraction method that suits a zone's shape.
type Technique string

const (
	// TechniqueContours suits continuous outdoor terrain.
	TechniqueContours Technique = "contours"
	// TechniqueSilhouette suits corridor and cave dungeons.
	TechniqueSilhouette Technique = "silhouette"
	// TechniqueBoundary suits structured interiors built from discrete floors.
	TechniqueBoundary Technique = "boundary"
)

// Classification records a zone's shape metrics and the technique chosen from
// them. The metrics are kept so the decision stays auditable rather than hidden
// behind a bare verdict.
type Classification struct {
	Zone           string
	Triangles      int
	WalkableFaces  int
	Occupancy      float64
	BoundaryDensit float64
	ZSpan          float64
	Technique      Technique
	// Overridden reports that the technique came from the manual table rather
	// than the thresholds.
	Overridden bool
}

// Classification thresholds, calibrated over all 178 Quarm zones (see
// docs/maps-feasibility.md 5b.6).
//
// Boundary density — unshared floor edges per floor face — is the primary
// signal because it is strongly bimodal: 71 zones cluster below 0.30, 87 above
// 0.60, with a sparse valley of 18 between. It measures the actual failure
// mode, since a continuous terrain mesh shares every interior edge.
//
// Occupancy is only a secondary split within the dense cluster. It looked
// bimodal on a five-zone sample but is roughly uniform across the full corpus,
// so it cannot carry the primary decision.
const (
	bdThreshold  = 0.45
	occThreshold = 0.60

	// sparseOccThreshold / sparseBDFloor catch dungeons and cave networks whose
	// boundary density lands in the terrain range.
	//
	// Established by rendering all 18 valley zones three ways: echo, powater,
	// netherbian, paludal, thedeep and warrens all took the terrain branch on
	// boundary density alone, and contours came out fragmentary or dashed on
	// every one of them while silhouette was clean. What they share is very low
	// occupancy — walkable area covers a quarter of the footprint or less, which
	// no open terrain zone does.
	//
	// The BD floor is what keeps fungusgrove on contours. It is the one
	// low-occupancy zone whose contours render continuous and informative, and
	// its density (0.24) sits well below the others (0.34-0.38) — a genuinely
	// more continuous mesh, not a corridor network.
	sparseOccThreshold = 0.35
	sparseBDFloor      = 0.30
)

// techniqueOverrides pins the zones no threshold gets right.
//
// Both are high-occupancy built structures whose boundary density happens to
// fall in the terrain range, so neither the sparse rule nor the density rule
// reaches them. Verified by rendering all three techniques: for each, contours
// are scattered fragments and the silhouette is a bare outline, while boundary
// extraction produces the real floor plan.
//
// warrens used to need an entry; the sparse rule now classifies it directly,
// which is the better outcome — a rule that generalises beats a pinned name.
var techniqueOverrides = map[string]Technique{
	// The arena ring, its alcoves and the entrance tunnel. occ=0.78, bnd=0.38.
	"arena": TechniqueBoundary,
	// Ssraeshza Temple's multi-level floor plan. occ=0.69, bnd=0.43.
	"ssratemple": TechniqueBoundary,
}

// Classify measures a zone and picks an extraction technique.
func Classify(z *Zone) Classification {
	c := Classification{Zone: z.ShortName, Triangles: len(z.Triangles)}

	faces := z.WalkableFaces()
	c.WalkableFaces = len(faces)
	if len(faces) == 0 {
		c.Technique = TechniqueBoundary
		return c
	}

	c.Occupancy = z.Rasterize().Occupancy()
	c.BoundaryDensit = float64(len(z.BoundaryEdges())) / float64(len(faces))

	minZ, maxZ := z.Vertices[0].Z, z.Vertices[0].Z
	for _, v := range z.Vertices {
		if v.Z < minZ {
			minZ = v.Z
		}
		if v.Z > maxZ {
			maxZ = v.Z
		}
	}
	c.ZSpan = maxZ - minZ

	switch {
	case c.Occupancy < sparseOccThreshold && c.BoundaryDensit >= sparseBDFloor:
		c.Technique = TechniqueSilhouette // sparse corridor / cave network
	case c.BoundaryDensit < bdThreshold:
		c.Technique = TechniqueContours // continuous terrain surface
	case c.Occupancy < occThreshold:
		c.Technique = TechniqueSilhouette
	default:
		c.Technique = TechniqueBoundary // discrete floor slabs
	}

	if t, ok := techniqueOverrides[z.ShortName]; ok {
		c.Technique, c.Overridden = t, true
	}
	return c
}

// contourInterval picks a slice spacing that keeps a zone's contour count
// readable regardless of how much vertical range it covers.
func contourInterval(zSpan float64) float64 {
	switch {
	case zSpan > 2000:
		return 100
	case zSpan > 1000:
		return 50
	case zSpan > 400:
		return 30
	default:
		return 20
	}
}

// contourRDPEpsilon simplifies contour lines. Contours come out of
// triangle-plane intersection as one segment per crossed triangle, which is far
// finer than anything visible: at 0.5 world units the error is well under a
// pixel at any usable zoom, and it roughly halves the segment count. Contours
// are the bulk of the corpus, so this is where size is won.
const contourRDPEpsilon = 0.5

// Extract runs the technique the classification selected and returns the map
// segments for the zone.
func Extract(z *Zone, c Classification) []Segment {
	switch c.Technique {
	case TechniqueContours:
		raw := z.Contours(contourInterval(c.ZSpan))
		// Chaining keys on Z, so separate elevation bands never join.
		return SimplifyRDP(Chain(raw), contourRDPEpsilon)
	case TechniqueSilhouette:
		return z.Silhouette()
	default:
		return SimplifyCollinear(Chain(z.BoundaryEdges()), 0.03)
	}
}

File: scripts/mapspike/classify.py
"""
Classify every Quarm zone by which extraction technique suits it.

Emits one row per zone with the metrics behind the decision, so the thresholds
stay auditable rather than hidden in a heuristic. See docs/maps-feasibility.md
section 5b.5.

    python3 classify.py            # all zones in quarm.db that have client files
    python3 classify.py unrest akheva
"""
import math, os, sqlite3, sys, csv, traceback
from collections import Counter

from mesh import zone_triangles
from silhouette import walkable_faces, rasterize
from floorplan3 import build

CLIENT = '/Volumes/T7/EQ/TAKPv22'
DB = '/Users/jasonsoprovich/repos/github.com/jasonsoprovich/pq-companion/backend/data/quarm.db'

OCC_THRESHOLD = 0.60      # below -> walkable area is sparse -> silhouette works
GAP_THRESHOLD = 0.25      # fraction of empty Z bins -> discrete floors present
BD_THRESHOLD  = 0.50      # unshared-edge density; below -> continuous terrain


def zone_list():
    db = sqlite3.connect(DB)
    names = [r[0] for r in db.execute("select distinct short_name from zone order by short_name")]
    return [z for z in names if os.path.exists(f'{CLIENT}/{z}.s3d')]


def z_structure(faces, binsize=20.0):
    """Return (z_span, gap_ratio). gap_ratio = fraction of bins between min and
    max Z that hold < 5% of the peak bin. Discrete floors leave empty air
    between them; ramps and terraces do not."""
    zs = [p[2] for f in faces for p in f]
    if not zs:
        return 0.0, 0.0
    lo, hi = min(zs), max(zs)
    if hi - lo < binsize:
        return hi - lo, 0.0
    h = Counter(int((v - lo) // binsize) for v in zs)
    nbins = int((hi - lo) // binsize) + 1
    peak = max(h.values())
    empty = sum(1 for b in range(nbins) if h.get(b, 0) < peak * 0.05)
    return hi - lo, empty / nbins


def classify(zone):
    verts, tris = zone_triangles(f'{CLIENT}/{zone}.s3d')
    faces = walkable_faces(zone)
    if not faces:
        return dict(zone=zone, tris=len(tris), faces=0, occupancy=0.0,
                    z_span=0.0, gap_ratio=0.0, bnd_density=0.0, technique='EMPTY')
    img, _ = rasterize(faces)
    occ = sum(1 for p in img.get_flattened_data() if p) / (img.size[0] * img.size[1])
    span, gap = z_structure(faces)

    # Boundary density = unshared floor edges per floor face. This measures the
    # actual failure mode for terrain: a continuous ground mesh shares every
    # interior edge, so boundary extraction yields almost nothing and contours
    # are the only workable technique. Discrete floor slabs score ~0.7-0.8.
    _, _, _, bnd = build(zone)
    bd = len(bnd) / len(faces)

    if bd < BD_THRESHOLD:
        tech = 'contours'          # continuous terrain surface
    elif occ < OCC_THRESHOLD:
        tech = 'silhouette'        # sparse walkable area -> corridors/caves
    else:
        tech = 'boundary'          # discrete floor slabs
    return dict(zone=zone, tris=len(tris), faces=len(faces), occupancy=occ,
                z_span=span, gap_ratio=gap, bnd_density=bd, technique=tech)


if __name__ == '__main__':
    zones = sys.argv[1:] or zone_list()
    rows, failed = [], []
    print(f"{'zone':<16}{'tris':>9}{'faces':>9}{'occ%':>7}{'z_span':>9}{'bnd_d':>7}  technique",
          flush=True)
    for z in zones:
        try:
            r = classify(z)
        except Exception as e:
            failed.append((z, f'{type(e).__name__}: {e}'))
            print(f"{z:<16}{'':>9}{'':>9}{'':>7}{'':>9}{'':>7}  FAILED  {type(e).__name__}",
                  flush=True)
            continue
        rows.append(r)
        print(f"{r['zone']:<16}{r['tris']:>9,}{r['faces']:>9,}{100*r['occupancy']:>6.1f}%"
              f"{r['z_span']:>9.0f}{r['bnd_density']:>7.2f}  {r['technique']}", flush=True)

    with open('zone_classification.csv', 'w', newline='') as fh:
        w = csv.DictWriter(fh, fieldnames=['zone','tris','faces','occupancy','z_span','gap_ratio','bnd_density','technique'])
        w.writeheader(); w.writerows(rows)

    print(f"\n=== {len(rows)} classified, {len(failed)} failed ===")
    for t, n in Counter(r['technique'] for r in rows).most_common():
        print(f"   {t:<12} {n:>4}  ({100*n/len(rows):.0f}%)")
    if failed:
        print("\nfailures:")
        for z, e in failed:
            print(f"   {z:<16} {e}")
